### PR TITLE
reset stream when receiving 101

### LINKED ISSUE #1017: Treat response code 101 as an error and reset the stream

### PR DESCRIPTION
**Pull request by KershawChang**
This will fix #1017

### CODE REVIEW

File: neqo-http3/src/connection_client.rs [hunk 1]
```diff
@@ -6728,4 +6728,42 @@ mod tests {
...
     }
+
+    #[test]
+    fn response_w_101() {
+        let (mut client, mut server, request_stream_id) = connect_and_send_request(true);
+
+        setup_server_side_encoder(&mut client, &mut server);
+
+        let mut d = Encoder::default();
+        let headers1xx: &[Header] = &[Header::new(":status", "101")];
```
Comment (martinthomson) [inline neqo-http3/src/connection_client.rs:6739]:
```suggestion
        let headers1xx = &[Header::new(":status", "101")];
```

File: neqo-http3/src/connection_client.rs [hunk 2]
```diff
@@ -6728,4 +6728,42 @@ mod tests {
...
+        server.encode_headers(request_stream_id, headers1xx, &mut d);
+
+        let headers200: &[Header] = &[
```
Comment (martinthomson) [inline neqo-http3/src/connection_client.rs:6742]:
```suggestion
        let headers200 = &[
```

Comment (martinthomson) [reply neqo-http3/src/connection_client.rs:6742]: Do you even need a 200 response?

File: neqo-http3/src/headers_checks.rs [hunk 3]
```diff
@@ -48,12 +48,12 @@ impl TryFrom<(MessageType, &str)> for PseudoHeaderState {
...
 /// Returns an error if response headers do not contain
 /// a status header or if the value of the header cannot be parsed.
-pub fn is_interim(headers: &[Header]) -> Res<bool> {
+pub fn is_interim(headers: &[Header]) -> Res<(bool, i32)> {
     let status = headers.iter().take(1).find(|h| h.name() == ":status");
     if let Some(h) = status {
         #[allow(clippy::map_err_ignore)]
         let status_code = h.value().parse::<i32>().map_err(|_| Error::InvalidHeader)?;
-        Ok((100..200).contains(&status_code))
+        Ok(((100..200).contains(&status_code), status_code))
```
Comment (martinthomson) [inline neqo-http3/src/headers_checks.rs:56]: Would it be possible to do this instead?

```suggestion
        if status_code == 101 {
            Err(Error::InvalidHeader)
        } else {
            Ok((100..200).contains(&status_code))
        }
```

That would integrate better into existing code.  It means generating an error when sending a message, but I think that's good too.

Comment (KershawChang) [reply neqo-http3/src/headers_checks.rs:56]: Right, doing this means less change for the current code.  Thanks.

File: neqo-http3/src/connection_client.rs (hunk 2 above)
Comment (KershawChang) [reply neqo-http3/src/connection_client.rs:6742]: Obviously, no. I'll remove this.

**Review by ddragana — APPROVED** [review]

**Pull request merged**